samples: Bluetooth: hci_rpmsg: Add vendor specific HCI events for BT assert and system fatal error

ppryga-nordic opened the pull request:
The PR adds two debugging enhancements to `hci_rpmsg` sample application:
1. Extended `bt_ctlr_assert_handle` to generate vendor specific HCI event with information about source and line where assert happened.
2.  Adds `k_sys_fatal_error_handler` to generate vendor specific HCI event with information about CPU state when a fault has happened. The handler is implemented for cortex-M CPUs only. It may be disabled by use of `VS_HCI_SYSTEM_FATAL_ERROR_HANDLER`

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

jhedberg commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
+		net_buf_add_u8(buf, 0x00);
+	}
+
+	net_buf_add_u8(buf, line >> 0 & 0xff);
+	net_buf_add_u8(buf, line >> 8 & 0xff);
+	net_buf_add_u8(buf, line >> 16 & 0xff);
+	net_buf_add_u8(buf, line >> 24 & 0xff);
```
Doesn't `net_buf_add_le32(buf, line);` do the same thing?

jhedberg commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
+	struct net_buf *buf;
+
+	buf = hci_debug_evt_create(BT_HCI_EVT_VENDOR, len);
+	if (!buf) {
+		BT_ASSERT_MSG(false, "Controller assert in: %s at %d", file, line);
+	}
```
`BT_ASSERT_MSG(msg != NULL, ...);` would be cleaner

jhedberg reviewed (changes requested):
See my couple of inline comments.

Additionally:

 - Isn't this something we might want to extend to other HCI transports as well?
 - What about the Zephyr HCI vendor extensions, would it make sense to specify and implement such an event in that context instead? That specification already includes a "Zephyr Fatal Error" event which might be extensible for including all the info you need: https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/connectivity/bluetooth/api/hci.txt#L1076

jhedberg commented:
> Isn't this something we might want to extend to other HCI transports as well?

Already got this partially answered by discovering #46295

ppryga commented:
@jhedberg 

> Isn't this something we might want to extend to other HCI transports as well?

Agree all other transports should have the same functionality.
I've added it to `hci_rpmsg` and `hci_uart` samples, because I've been using this samples. I understand that you suggest to add it to: `hci_spi`, `hci_usb` and `hci_usb_h4`, correct? 

> What about the Zephyr HCI vendor extensions, would it make sense to specify and implement such an event in that context instead? That specification already includes a "Zephyr Fatal Error" event which might be extensible for including all the info you need: https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/connectivity/bluetooth/api/hci.txt#L1076

This is something I was not aware of. We have a `Zephyr Fatal Error` event then I'll try to extend it and use instead of implicitly create a new kind of an event in sample code.

Thalley commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
+#if defined(CONFIG_VS_HCI_SYSTEM_FATAL_ERROR_HANDLER)
+void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *esf)
+{
+	/* 0xAA + r0/a1 + 0xAA + r1/a2 + 0xAA + r2/a3 + 0xAA + r3/a4 + 0xAA + r12/ip +
+	 * 0xAA + r14/lp + 0xAA + XPSR
+	 */
+	uint32_t len = 7 * (1 + 4);
```
See my comments in https://github.com/zephyrproject-rtos/zephyr/pull/46295

Thalley commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
+	/* Send the event over rpmsg */
+	hci_rpmsg_send(buf);
+
+	while (1) {
+	}
```
`while (true);`?

ppryga replied:
It fails with checkpatch:
```
ERROR:TRAILING_STATEMENTS: trailing statements should be on next line
#184: FILE: samples/bluetooth/hci_rpmsg/src/main.c:301:
+	while (true);
```

jhedberg commented:
> I've added it to hci_rpmsg and hci_uart samples, because I've been using this samples. I understand that you suggest to add it to: hci_spi, hci_usb and hci_usb_h4, correct?

Yes, but it might then make sense to factor out the common parts to a single place instead of having duplicated code everywhere.

ppryga commented:
@jhedberg 
> Yes, but it might then make sense to factor out the common parts to a single place instead of having duplicated code everywhere.

Yeah, I have to look around for a best place where to put it. It would be the best that any application has access to that functionality.

cvinayak commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
 	} while (ret < 0);
 
 	LOG_INF("Sent message of %d bytes.", ret);
+	printk("Sent message of %d bytes.", ret);
```
Any reason to use both logging and printk?

cvinayak commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
 }
 
+#if defined(CONFIG_BT_CTLR_ASSERT_HANDLER) || defined(CONFIG_VS_HCI_SYSTEM_FATAL_ERROR_HANDLER)
+/* A memory pool debug purposes. Currently used sent HCI vendor specific event with LL_ASSERT
+ * information.
+ */
```
```suggestion
/* A memory pool for debug purposes. Currently used for sending HCI vendor specific event with LL_ASSERT
 * information.
 */
```

cvinayak commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
+	/* Disable interrupts, this is unrecoverable */
+	(void)irq_lock();
+
+	if (file) {
+		while (file[len] != '\0') {
+			if (file[len] == '/') {
+				pos = len + 1;
+			}
+			len++;
+		}
+		file += pos;
+		len -= pos;
+	}
```
Add a comment as to what the loop is trying to do.

cvinayak commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
+
+	buf = hci_debug_evt_create(BT_HCI_EVT_VENDOR, len);
+	if (!buf) {
+		BT_ASSERT_MSG(false, "Controller assert in: %s at %d", file, line);
```
Could this be recursive, being called inside assert handler?

cvinayak commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
+
+		buf = hci_debug_evt_create(BT_HCI_EVT_VENDOR, len);
+		if (!buf) {
+			BT_ASSERT_MSG(false, "Fatal error\n");
```
Same here, will this get recursive?

cvinayak commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
+#if defined(CONFIG_VS_HCI_SYSTEM_FATAL_ERROR_HANDLER)
+void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *esf)
+{
+	/* 0xAA + r0/a1 + 0xAA + r1/a2 + 0xAA + r2/a3 + 0xAA + r3/a4 + 0xAA + r12/ip +
+	 * 0xAA + r14/lp + 0xAA + XPSR
+	 */
```
Why do you need `0xAA` before every register value?  Use `0x55` or something else a new subevent code for the vendor specific HCI event.

cvinayak commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
+		BT_ASSERT_MSG(false, "Controller assert in: %s at %d", file, line);
+	}
+
+	net_buf_add_u8(buf, 0xAA);
```
```suggestion
	/* HCI vendor specific event subevent code
	 * asserting filename and line number as 
	 * parameters.
	 */
	net_buf_add_u8(buf, 0xAA);
```

cvinayak commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
+			BT_ASSERT_MSG(false, "Fatal error\n");
+		}
+
+		net_buf_add_u8(buf, 0xAA);
```
```suggestion
		/* HCI vendor specific subevent code with
		 * exception stack frame values as parameters.
		 */
		net_buf_add_u8(buf, 0x55);

```

cvinayak commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
+		}
+
+		net_buf_add_u8(buf, 0xAA);
+		net_buf_add_le32(buf, esf->basic.a1);
+
+		net_buf_add_u8(buf, 0xAA);
+		net_buf_add_le32(buf, esf->basic.a2);
+
+		net_buf_add_u8(buf, 0xAA);
+		net_buf_add_le32(buf, esf->basic.a3);
+
+		net_buf_add_u8(buf, 0xAA);
+		net_buf_add_le32(buf, esf->basic.a4);
+
+		net_buf_add_u8(buf, 0xAA);
+		net_buf_add_le32(buf, esf->basic.ip);
+
+		net_buf_add_u8(buf, 0xAA);
+		net_buf_add_le32(buf, esf->basic.lr);
+
+		net_buf_add_u8(buf, 0xAA);
+		net_buf_add_le32(buf, esf->basic.xpsr);
```
```suggestion
		net_buf_add_le32(buf, esf->basic.a1);
		net_buf_add_le32(buf, esf->basic.a2);
		net_buf_add_le32(buf, esf->basic.a3);
		net_buf_add_le32(buf, esf->basic.a4);
		net_buf_add_le32(buf, esf->basic.ip);
		net_buf_add_le32(buf, esf->basic.lr);
		net_buf_add_le32(buf, esf->basic.xpsr);
```

Thalley replied:
This will not give the same, right? 

The original code will be something like
a1, 0xAA, a2, 0xAA, a3, ...
where what you suggest will give
a1, a2, a3

ppryga replied:
Well, I was about to change the implementation to use something that was proposed by @jhedberg:
https://github.com/zephyrproject-rtos/zephyr/blob/360d70a91ea731d47f3a7f84c154f397213f18e6/doc/connectivity/bluetooth/api/hci.txt#L1076
The `fatal_error` is aimed for general fatal errors. It provides PC register and some string with error description e.g. source file and line number. 

Though, it will be an architecture specific, because `z_arch_efs_t` is an alias and its real definition is not common for all architectures. For Cortex-M it will hold: `A1`, `A2`, `A2`, `A3`, `A4`, `IP`, `LR`, XPSR` registers. 

```
#define BT_HCI_EVT_VS_SYS_FATAL_ERROR 0x5
struct bt_hci_evt_vs_sys_fatal_error {
	uint32_t reason;
#if defined(CONFIG_CPU_CORTEX_M)
	uint32_t a1;
	uint32_t a2;
	uint32_t a3;
	uint32_t a4;
	uint32_t ip;
	uint32_t lr;
	uint32_t xpsr;
#endif /* CONFIG_CPU_CORTEX_M */
};
```

I've created a new event, because it seems to me that I can't change the existing one `bt_hci_evt_vs_fatal_error`. It is part of the Zephyrs HCI API that is already released. It will break backward compatibility.

BTW I didn't find any use of `bt_hci_evt_vs_fatal_error` in the Controller. The other thing is that, it is called a Zephyr Fatal Error and description says:
> This event reports a fatal non-recoverable error from the local Controller.

I understand it as a Controller only error. Not system wide, non-recoverable error.

ppryga replied:
@Thalley I was not sure about it, so I wrote my last message to get an input. If there is no problem with change of the `bt_hci_evt_vs_fatal_error` then I'll do that and make it to be system wide fatal error HCI event. To asserts and other Controller only errors I suggest to create a new VS HCI event.

@cvinayak what it your opinion?

ppryga replied:
Well I thought about following distinction:
- `Zephyr Fatal Error` e.g. hard fault, kernel error, etc. anything that is unrecoverable and requires a system/core to be reset
- `Controller Fatal Error` e.g. Controller's LL_ASSERT. That does not mean the core/system has to be restarted, but Controller only. Other errors may fall into that category in future.

For start `Zephyr Fatal Error` would be implemented for Cortex-M. It would include stack frame information like one that is printed by Zephyr in such error happens. The `Controller Fatal Error` would encapsulate information where LL_ASSERT happened: source file and line.

Correct me if I'm wrong in my understanding of the Host view point. Since Host communicates with Controller through a defined HCI interface it does not matter if Controller does not response because of its internal error or system errors. It is just a not responding Controller. 

The thing is a bit different if you take into account conformance tests. While debugging it does matter if the issue is an `LL_ASSERT` or `Zephyr Fatal Error`. Debugging conformance tests is the original reason why I've created these two PRs.

ppryga replied:
Hello all,

Is the current approach okay for you? If it is, then I'll prepare commits to apply same solution for other `hci_XXX1` samples also.

cvinayak commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -235,17 +239,146 @@ static void hci_rpmsg_send(struct net_buf *buf)
+		hci_rpmsg_send(buf);
+	}
+
+	LOG_PANIC();
+	LOG_ERR("Halting system");
```
Will this work with irq disabled? Is it to get logs out on RTT/UART console?

jori-nordic reviewed (dismissed):
Nothing to say apart from the comments already there.

ppryga commented:
I've added a commit with a proposal how to change/extend Zephyr Fatal Error and use in by sample applications.
Have a look please. Pay attention that code needs to be polished. I didn't want to spend too much time on it being just a proposal.

Looking forward for your comments.

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4798,6 +4798,95 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+		len -= pos;
+	}
+
+	/* Total length: fine name strlen + \0 + 32-bit line number */
+	len = 1 + len + 1 + 4;
```
So the last 1 and the 4 is the `\0` and the 32-bit line number, and I guess the `len` is the name (do you mean `file` and not `fine`?), but what is the first `1`?

ppryga replied:
Yes it should be `file` and first `1` is a leftover.

ppryga replied:
Corrected

Thalley commented on include/zephyr/bluetooth/hci_vs.h:
```diff
@@ -216,9 +216,25 @@ struct bt_hci_evt_vs {
+	uint32_t lr;
+	uint32_t xpsr;
+#endif /* CONFIG_CPU_CORTEX_M */
+};
```
Should be `__packed`

ppryga replied:
Corrected

Thalley commented on include/zephyr/bluetooth/hci_vs.h:
```diff
@@ -216,9 +216,25 @@ struct bt_hci_evt_vs {
+#define BT_HCI_EVT_VS_ERROR_TYPE_CTRL_ASSERT 0x2
+struct bt_hci_vs_fatal_error_stack_frame {
+	uint32_t reason;
+#if defined(CONFIG_CPU_CORTEX_M)
+	uint32_t a1;
+	uint32_t a2;
+	uint32_t a3;
+	uint32_t a4;
+	uint32_t ip;
+	uint32_t lr;
+	uint32_t xpsr;
+#endif /* CONFIG_CPU_CORTEX_M */
```
Not really sure how I feel about having a HCI event (even VS) to be have CPU type dependencies like this, but I guess it makes sense. 

Is this struct placed as the `data` in `bt_hci_evt_vs_fatal_error`? 

Would it make sense to already now introduce it as a `union`, so that it is easier to extend to other CPU types?

ppryga replied:
> Is this struct placed as the data in bt_hci_evt_vs_fatal_error?

Yes it is used for `system fatal error` case as a `data` added to event.

>Would it make sense to already now introduce it as a union, so that it is easier to extend to other CPU types?

I think that is a good idea. For `raw` bytes/data like for a `string` user can go with `data` member, for other cases named member shall be provided.

Thalley replied:
Since we already have the `type` defined, I guess we could use that to determine which struct to cast the data to, right? 

Does the data need to CPU specific, or do we just need to supply different formats, based on the type field? That would be similar to how HCI is defined in the first place as LTV structures.

ppryga replied:
There are two options:
- we consider Host as being aware of an architecture of CPU where Controller runs
- or I'll add another common field to `bt_hci_evt_vs_fatal_error` with information about architecture. The CPU specific part may be also a union:
```
union {
        uint8_t cpu_data[0];
        struct bt_hci_evt_err_stack_frame sf_cortex_m;
};
```

I think that `cpu_data_length` is not required. The total length is already available in the event. The length of the data filed may be inferred from `type` and if available `cpu_type`.

ppryga replied:
> Would it make sense to already now introduce it as a union, so that it is easier to extend to other CPU types?
and
> or I'll add another common field to `bt_hci_evt_vs_fatal_error` with information about architecture. The CPU specific part may be also a union

I do not want to create a unions there. If there is added support for another architecture that uses more data than Cortex-M stack frame, then the size of for Cortex-M will increase also. Additional bytes will be not used and they have to be zeroed anyway. I think it is better to construct the `data` and `cpu_data` by copying and have smallest possible event.

Thalley replied:
> > 
> 
> I do not want to create a unions. If there is added support for another architecture that uses more data than Cortex-M stack frame, then the size of for Cortex-M will increase also. Additional bytes will be not used and they have to be zeroed anyway. I think it is better to construct the `data` and `cpu_data` by copying and have smallest possible event.

If we guard the union proper with e.g. `#if defined(CONFIG_CPU_CORTEX_M)`, then the size shouldn't be an issue. 

The point of defining the union was to add a reference to what structs could be put in `data`, e.g.

```c
struct bt_hci_vs_fata_error_cpu_data_cortex_m {
	uint32_t a1;
	uint32_t a2;
	uint32_t a3;
	uint32_t a4;
	uint32_t ip;
	uint32_t lr;
	uint32_t xpsr;
} __packed;

struct bt_hci_vs_fata_error_cpu_data_cortex_x {
	uint32_t a1;
	uint32_t a2;
	uint32_t a3;
	uint32_t a4;
} __packed;

union bt_hci_evt_vs_fatal_error_data {
#if defined(CONFIG_CPU_CORTEX_M)
	struct bt_hci_vs_fata_error_cpu_data_cortex_m cortex_m.
#endif /* CONFIG_CPU_CORTEX_M */
#if defined(CONFIG_CPU_CORTEX_X)
	struct bt_hci_vs_fata_error_cpu_data_cortex_x cortex_x.
#endif /* CONFIG_CPU_CORTEX_X */
};

struct bt_hci_evt_vs_fatal_error {
	uint8_t type;
	union bt_hci_evt_vs_fatal_error_data data;
} __packed;
```

where `CORTEX_X` is some other CPU.

ppryga replied:
@Thalley that is fine from Controller perspective because it runs on a particular architecture. 
It is also fine from a conformance tests perspective because then we just read the received HCI events and interpret them.
However looking from Host perspective, I'm not sure if this is how it should be defined. Host may be not aware of architecture of a CPU the Controller is running on. Then the `cpu_data` should be interpreted dynamically. 
If we go that path, then in my humble opinion `bt_hci_evt_vs_fatal_error_data data;` should not be a union.

Thalley replied:
That is a good point, however since this is an event, the host will never need to allocate it right? So the size of the `struct bt_hci_evt_vs_fatal_error` should not really matter for the host. 

I am fine with just keeping the `data` as a non-union though.

Thalley replied:
Actually, since the host CPU may be different from the controller CPU, the design with the guards based on CPU type won't work. Please ignore my suggestion :D

ppryga commented:
Hello all,

Is the current approach okay for you? If it is, then I'll prepare commits to apply same solution for other `hci_XXX` samples also.

Thalley commented on include/zephyr/bluetooth/hci_vs.h:
```diff
@@ -216,9 +216,29 @@ struct bt_hci_evt_vs {
 struct bt_hci_evt_vs_fatal_error {
-	uint64_t pc;
-	uint8_t  err_info[0];
+	uint8_t type;
+	uint8_t data[0];
```
I realize that we haven't done much documentation in this file, but it is a public header file that are occasionally used by applications. Should we consider documenting this a bit? At least I don't think it's clear to everyone later that the `data` can be a `struct bt_hci_vs_fatal_error_stack_frame` and that `cpu_data` can be a `struct bt_hci_vs_fata_error_cpu_data_cortex_m`.

ppryga replied:
For sure I'll add more extensive documentation. I didn't want to spend time on that in case the whole approach to the topic is not accepted.

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4798,6 +4798,111 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+	}
+
+	/* Total length: file name strlen + \0 + 32-bit line number */
+	len = len + 1 + 4;
```
```suggestion
	len = len + 1 + sizeof(line);
```

Thalley reviewed (commented):
I think the approach is OK

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4798,6 +4798,136 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+
+	if (file) {
+		while (file[len] != '\0') {
+			if (file[len] == '/') {
+				pos = len + 1;
+			}
```
Can you elaborate on why we are updating `pos` on a `/`?

ppryga replied:
Ah, yes that is not clear I'll provide a comment about it.
The `pos` is used because here the file name is extracted from path. 
From `gcc` documentation: 
> __FILE__
This macro expands to the name of the current input file, in the form of a C string constant. **This is the path by which the preprocessor opened the file**, not the short name specified in ‘#include’ or as the input file name argument. For example, "/usr/local/include/myheader.h" is a possible expansion of this macro.

Thalley replied:
Ah, so `file` is the _file path_?

ppryga replied:
yes and it would be better to change a name of the argument to `file_path`.

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4798,6 +4798,138 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+#if defined(CONFIG_CPU_CORTEX_M)
+typedef struct bt_hci_vs_fata_error_cpu_data_cortex_m bt_hci_vs_fatal_error_cpu_data;
+
+static void vs_err_fatal_cpu_data_fill(bt_hci_vs_fatal_error_cpu_data *cpu_data,
```
Hmm, is `bt_hci_vs_fatal_error_cpu_data` a struct? Looks like the name of a variable instead.

ppryga replied:
It is actually a `typedef`. The name was chosen to follow convention that I found for vendor specific types/structures in source code. Here are some examples: `bt_hci_cp_vs_read_tx_power_level`, `bt_hci_cp_vs_set_min_num_used_chans`.

Have you got a suggestion how to name it?

Thalley replied:
Oh sorry, I had simply mistaken the typedef for a variable >.> 

You can ignore my comment :)

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4798,6 +4798,138 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+
+	if (file_path) {
+		/* Extract file name from a path */
+		while (file_path[len] != '\0') {
+			if (file_path[len] == '/') {
+				pos = len + 1;
+			}
+			len++;
+		}
```
Minor: It may be more efficient to start from the end and stop at the first `/` to extract the file name

ppryga replied:
That would be more efficient but to achieve I would have to have a length of the `file_path`. That would require to go at least once over the whole string to find `'\0'`. So the optimization is lost.

That would be possible if `bt_ctlr_assert_handle` and `LL_ASSERT` implementation was changed. Unfortunately that change is a bit out of scope of this PR.

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4798,6 +4798,138 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+		file_path += pos;
+		len -= pos;
+
+		if (len) {
```
if `len == 0` does that mean no file name was found?

ppryga replied:
Yes, it can happen in two cases: `file_path` is empty string or `file_path` has following form `foo/bar/`.

I'll add a comment to avoid confusion.

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4798,6 +4798,138 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+			}
+
+			/* Prepare vendor specific HCI Fatal Error event */
+			buf = vs_err_evt_create(data_type, len);
```
`len` is a `uint32_t` but `vs_err_evt_create` takes a `uint8_t`. Should we verify that `len < UINT8_MAX` before creating the event?

ppryga replied:
Yes, good catch. Thank you!

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4798,6 +4798,138 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+				net_buf_add_le64(buf, pc);
+			}
+
+			while (*file_path != '\0') {
+				net_buf_add_u8(buf, *file_path);
+				file_path++;
+			}
```
```suggestion
			net_buf_add_mem(buf, file_path, strlen(file_path);
```

ppryga replied:
I'll introduce new variable `data_len` to store total event data length. Then new `strlen(file_path)` is not needed, I'll use `len` instead: `net_buf_add_mem(buf, file_path, len);`

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4798,6 +4798,138 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+				net_buf_add_u8(buf, *file_path);
+				file_path++;
+			}
+			net_buf_add_u8(buf, 0x00);
```
Null terminator?

ppryga replied:
Yes, I'll change 0x00 to a `#define`

ppryga commented:
Hi all, I've added missing documentation and did changes in `hci_rpmsg` `main.c` and `hci.c` files to avoid issues with use of `BT_ASSERT)MSG` and `k_yield`.

ppryga commented:
@Thalley comments were addressed

Thalley commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -33,13 +38,19 @@ static K_THREAD_STACK_DEFINE(tx_thread_stack, CONFIG_BT_HCI_TX_STACK_SIZE);
 static K_FIFO_DEFINE(tx_queue);
 static K_SEM_DEFINE(ipc_bound_sem, 0, 1);
+#if defined(CONFIG_BT_CTLR_ASSERT_HANDLER) || defined(CONFIG_BT_HCI_VS_FATAL_ERROR)
+static bool ipc_ept_ready;
```
And since this is a sample, may be worth to add a comment about this too (we have received feedback on Discord that many samples are using acronyms and abbreviations that aren't obvious to many new users).

ppryga replied:
I'll add a comment. Regarding the name, it follows convention from this source file e.g. 'hci_ept_bound`, `hci_ept_recv`.
It looks like we have two conventions for end point naming.

Thalley commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -230,7 +241,11 @@ static void hci_rpmsg_send(struct net_buf *buf)
 				retries = 0;
 			}
-			k_yield();
+			if (is_fatal_err) {
+				LOG_ERR("IPC service send error: %d", ret);
+			} else {
+				k_yield();
+			}
```
Can you elaborate on why we are not `k_yield`ing when sending fatal errors?

ppryga replied:
In case function `hci_rpmsg_send` is executed in a fatal error context, e.g. by `k_sys_fatal_error_handler` the whole system is in unstable state. There is no point to call `k_yield` in this case because it could make more harm. We do not know if there is any other thread running or if other kernel object are in a valid state. 

I've added a condition here because I didn't want to implement special version of the function for fatal error purposed.

Thalley replied:
Can you add that reasoning as a comment in the code as well?

ppryga replied:
I've added an explanation in the code comment.

Thalley commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -242,13 +257,75 @@ static void hci_rpmsg_send(struct net_buf *buf)
 static void hci_ept_bound(void *priv)
 {
 	k_sem_give(&ipc_bound_sem);
+#if defined(CONFIG_BT_CTLR_ASSERT_HANDLER) ||                                  \
+	defined(CONFIG_BT_HCI_VS_FATAL_ERROR)
```
On line 41 you have these on a single line:
```c
#if defined(CONFIG_BT_CTLR_ASSERT_HANDLER) || defined(CONFIG_BT_HCI_VS_FATAL_ERROR)
```

ppryga replied:
Aligned with line 41.

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4798,6 +4800,147 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+			(bt_hci_vs_fatal_error_cpu_data *)sf->cpu_data, esf);
+	} else {
+		BT_ERR("Can't create HCI Fatal Error event");
+
```
```suggestion
```

ppryga replied:
Removed

ppryga replied:
Removed

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4798,6 +4800,147 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+				} else {
+					BT_ERR("Can't create HCI Fatal Error event");
+				}
+			} else {
+				BT_ERR("Too much data for a HCI Fatal Error event");
+			}
```
Could we simply truncate the filename? Or does it need to be complete for the event to properly work?

ppryga replied:
Thank for a good tip, the filename may be truncated. I thing that file name with length around 200 characters is unique enough that it can be truncated. 

BTW the actual max length should be `BT_BUF_RX_SIZE` not `UINT8_MAX`. I'll use the `BT_BUF_RX_SIZE` in the code.

ppryga-nordic committed:
Bluetooth: hci: Extend VS fatal error for hard hault and assert

Extend current implementation of vendor-specific HCI events related
with Zephyr Fatal Error. The extended event is able to carry Zephyr
system wide fatal error as well as Controller related assertion.
That gives a possibility to use the event by bt_ctlr_assert_handle
or k_sys_fatal_error_handler to send information about unrecoverable
failures over HCI interface.

There are added two types of errors to Zephyr Fatal Error event:
- system fault error like e.g. Hard Fault. This one caries reason of
  a fault and a stack frame in the data field. The stack frame is
  defined for Corte-M SoCs only.
- Controller assert. The data field caries file name and line of code
  where an assert happened.

Except that the extended event can be used by Host, it can serve as a
debug mean to conformance testing of the Controller.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: hci: Use extended VS fatal error in hci and hci_rpmsg sample

Provide common helper functions to create extended extended Zephyr
Fatal Error functionality in HCI common code.

Use the implementation in hci_rpmsg sample.

The sample didn't provide an information about Controllers assert
or system fatal error to an application code while run with nRF5340
SoC. The goal for hci_rpmsg sample change is to enhance user experience
for conformance testing of the Bluetooth Controller while executed with
nRF5340.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
New @Thalley comments addressed.
@jhedberg the PR is split according to suggestion.

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4798,6 +4800,151 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+
+	buf = vs_err_evt_create(BT_HCI_EVT_VS_ERROR_DATA_TYPE_STACK_FRAME,
+				sizeof(*sf) + sizeof(*cpu_data));
+	if(buf != NULL) {
```
```suggestion
	if (buf != NULL) {
```

ppryga replied:
Corrected

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4798,6 +4800,151 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+			}
+
+			if (data_len + len > BT_BUF_RX_SIZE) {
+				uint32_t overflow_len = len + data_len - BT_BUF_RX_SIZE;
```
Shouldn't it actually be `BT_BUF_EVT_RX_SIZE`?

ppryga replied:
No, `BT_BUF_EVT_RX_SIZE` is used in buffers memory definition:
```
NET_BUF_POOL_FIXED_DEFINE(vs_err_tx_pool, 1, BT_BUF_RX_SIZE, 8, NULL);
```
`BT_BUF_EVT_RX_SIZE` may equal `BT_BUF_RX_SIZE` if `BT_BUF_EVT_RX_SIZE` is the biggest one of all other kinds of buffers used in HCI.

Thalley replied:
Right, but if `BT_BUF_EVT_RX_SIZE` is e.g. 150, and `BT_BUF_ACL_RX_SIZE` is 200, then `BT_BUF_RX_SIZE` is 200, but how can we then create an event here of size (`BT_BUF_RX_SIZE`/200) if the host uses `BT_BUF_EVT_RX_SIZE`/150 for the event pool?

```c
NET_BUF_POOL_FIXED_DEFINE(evt_pool, CONFIG_BT_BUF_EVT_RX_COUNT,
			  BT_BUF_EVT_RX_SIZE, 8,
			  NULL);
```

Wouldn't events > 150 get discarded then?

ppryga replied:
Well Host cannot use an event buffer smaller than 255 bytes if it is not 100% that all events fit into that size. This is an optimization for cooperation with well known Controller. 
https://github.com/zephyrproject-rtos/zephyr/blob/0cb22dfa1de5be2b828d67a9748b157105e3ba43/subsys/bluetooth/host/hci_core.c#L119-L125
and:
https://github.com/zephyrproject-rtos/zephyr/blob/d96f361175ba759d7344edbea3a3b5f1e766cba8/include/zephyr/bluetooth/buf.h#L90-L92
Nevertheless for optimization purpose I've changed the `evt_pool` definition to use `BT_BUF_EVT_RX_SIZE` because it is used only for events. That can save couple of bytes in case the event buffer size is not maximum allowed.

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4798,6 +4800,151 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
+				uint32_t overflow_len = len + data_len - BT_BUF_RX_SIZE;
+
+				/* Truncate the file name length by number of overflow bytes */
+				data_len = data_len + (len - overflow_len);
```
Isn't `len` the length of the file name (perhaps we need to rename it to be more explicit?)? 

You are still doing `net_buf_add_mem(buf, file_path, len);`, and since we are not reducing `len` here, it seems that we are still adding the entire file path (do we need the file path, and not just the file name?).

ppryga replied:
I thought about rename it, but decided to leave it as is... but it would be better to have an explicit name like `file_name_len`. I'll update it.

`net_buf_add_mem(buf, file_path, file_name_len);` is correct, because the `file_name_len` if updated to store correct file name length after end of `while` loop that searches for the file name:
```
/* Extract file name from a path */
while (file_path[file_name_len] != '\0') {
	if (file_path[file_name_len] == '/') {
		pos = file_name_len + 1;
	}
	file_name_len++;
}
file_path += pos;
file_name_len -= pos; // <--------- here full path length is updated to store file name length. pos stores position of last '/' + 1
```

ppryga replied:
You were right, there were an issue with `file_name_len` it was not subtracted with `overflow_len` before `net_buf_add_mem(buf, file_path, file_name_len);` 
What more, the `data_len` was updated with `file_name_len` only in case of overflow. The `data_len += file_name_len;` should be done after `if` statement that checks the overflow.

Both issues are fixed now.

Thalley reviewed (commented):
LGTM, but still got one outstanding comment that needs to be answered first

ppryga commented:
Changed the `evt_pool` definition to use `BT_BUF_EVT_RX_SIZE` instead of `BT_BUF_RX_SIZE` for memory optimization.

ppryga commented:
I've fixed compliance issues except new `typedef`. Can anyone say if is it allowed to create new typedefs, no matter the checkpatch complains about it?

Thalley commented:
> I've fixed compliance issues except new `typedef`. Can anyone say if is it allowed to create new typedefs, no matter the checkpatch complains about it?

I guess the reason why it's complaining is that it's opposing the Linux Style guide: https://www.kernel.org/doc/html/v4.10/process/coding-style.html#typedefs

ppryga commented:
>> I've fixed compliance issues except new typedef. Can anyone say if is it allowed to create new typedefs, no matter the checkpatch complains about it?

>I guess the reason why it's complaining is that it's opposing the Linux Style guide: https://www.kernel.org/doc/html/v4.10/process/coding-style.html#typedefs

Actually it falls into:
> a. totally opaque objects (where the typedef is actively used to hide what the object is).
from Linux Style guide.

The goal for the `typedef` is to make the `bt_hci_vs_fatal_error_cpu_data` opaque for `hci_vs_err_stack_frame` function. In case there is another architecture that uses the same functionality, then `vs_err_fatal_cpu_data_fill` will be "overloaded" (guarded by Kconfigs) and different type provided under `bt_hci_vs_fatal_error_cpu_data`. Thanks to that we have single implementation of `hci_vs_err_stack_frame`. 

In my humble opinion it falls into mentioned point from Linux Style guide. @Thalley what do you think?

Thalley commented:
> > > I've fixed compliance issues except new typedef. Can anyone say if is it allowed to create new typedefs, no matter the checkpatch complains about it?
> 
> > I guess the reason why it's complaining is that it's opposing the Linux Style guide: https://www.kernel.org/doc/html/v4.10/process/coding-style.html#typedefs
> 
> Actually it falls into:
> 
> > a. totally opaque objects (where the typedef is actively used to hide what the object is).
> > from Linux Style guide.
> 
> The goal for the `typedef` is to make the `bt_hci_vs_fatal_error_cpu_data` opaque for `hci_vs_err_stack_frame` function. In case there is another architecture that uses the same functionality, then `vs_err_fatal_cpu_data_fill` will be "overloaded" (guarded by Kconfigs) and different type provided under `bt_hci_vs_fatal_error_cpu_data`. Thanks to that we have single implementation of `hci_vs_err_stack_frame`.
> 
> In my humble opinion it falls into mentioned point from Linux Style guide. @Thalley what do you think?

Given that the typedef only has a single use at this point, it's not as clear as to why it's needed/useful compared to simply using `struct bt_hci_vs_fata_error_cpu_data_cortex_m`, but I assume that it's going to be used for non-cortex-m cpu data, right? 

I'm OK with it, but I do not have the final say here I think :D

ppryga commented:
> Given that the typedef only has a single use at this point, it's not as clear as to why it's needed/useful compared to simply using struct bt_hci_vs_fata_error_cpu_data_cortex_m, but I assume that it's going to be used for non-cortex-m cpu data, right?

> I'm OK with it, but I do not have the final say here I think :D

I did it because I found it better to prepare for possible other future implementations. It that is not welcome then I can remove it and use `bt_hci_vs_fata_error_cpu_data_cortex_m` directly.

ppryga commented:
@jhedberg @cvinayak can you re-review the PR?

ppryga commented:
@cvinayak @jhedberg your request for changes blocks the PR. Will you be able to re-review it?

cvinayak commented on doc/connectivity/bluetooth/api/hci.txt:
```diff
@@ -1077,25 +1077,124 @@ Zephyr Fatal Error
+	| a3                 | 4 octets | General purpose register        |
+	| a4                 | 4 octets | General purpose register        |
+	| ip                 | 4 octets | Instruction pointer register    |
+	| a1                 | 4 octets | Link register                   |
```
```suggestion
	| lr                 | 4 octets | Link register                   |
```

ppryga replied:
Corrected

ppryga commented:
@jhedberg can you re-review the PR? Or please withdraw you change request in other case.

jhedberg reviewed (approved):
Sorry for the delay, just got back from summer vacation and catching up on PR reviews.

carlescufi commented:
Overriding compliance, which fails due to a new `typedef` being introduced.
